### DISCUSSION TITLE
Anyone else unable to connect to HA with "Supervisor token not found" errors? Killed all my automations :-(

### DISCUSSION THREAD

**Discussion by lowgatelabs** [zachowj/hass-node-red#145] [Q&A]
I've been using Node-RED running in a docker, and HA in its standard VM, happily talking via node-red companion for at least a year.

My Node-RED updated to Version 3.0, and now my HA-related nodes are unable to connect and on NR startup I get these three errors:

6 Aug 19:52:06 - [error] [server:Home Assistant Connect] Supervisor token not found.
6 Aug 19:52:06 - [error] [webhookrelay:76652f56.8700a] No Webhook Relay token key set
6 Aug 19:52:06 - [error] [webhookrelay:76652f56.8700a] No Webhook Relay token secret set

On the HA side, I'll get:

2022-08-06 14:51:35.432 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [139999888081344] Error handling message: Unauthorized (unauthorized)

Clearly this is some sort of login problem and some token gets generated somewhere, but I can't figure out where or how to fix.

What's odd is that if I change the HA configuration in a node red icon and UNCHECK "Using the Home Assistant Add-on" and populate a key, I can connect to HA, although this breaks some of my automations since the entity names seem to differ slightly when using the Add-On vs. not using the add on.

If I think recheck "Use the Add-On" it seems to work for some period.

Things I've tried:

1. Restart HA and Node-Red in various orders/combinations
2. Redownloaded the Companion in HACS (although I haven't tried deleting and re-installing) followed by various reboots
3. Creating a "credentialSecret" in my Node Red configuration based on another thread that seemed to show a similar problem
4. Did a rain dance, sacrificed a chicken, and said "Candyman" three times

Any ideas would be greatly appreciated since 100% of my automation (and there's a fair amount) is in NR. Thanks!!!

**Comment by zachowj** [comment]
If you're running the NR add-on, it sets an environment variable `SUPERVISOR_TOKEN` which the NR HA nodes use to connect to the add-on proxy. The "Supervisor token not found" error is thrown when the node can't find that environment variable. If this continues to happen after a reboot of the NR add-on you might want to ask in the add-on repo.

 > I can connect to HA, although this breaks some of my automations since the entity names seem to differ slightly when using the Add-On vs. not using the add on.

If you connect to HA using the add-on proxy or directly your entity ids are not going to be different.

> If I think recheck "Use the Add-On" it seems to work for some period.

It works for "some period"? What stops working? What do your NR logs show?

**Reply by lowgatelabs** [reply]
It appears I was wrong about the "staying connected for some period." It looked like the nodes were connected in the UI since they still had status underneath, but as soon as I ticked the "Use the addon" option I get the following in my Node Red logs:

6 Aug 21:22:41 - [error] [server:Home Assistant Connect] Supervisor token not found.
6 Aug 21:22:41 - [info] Started modified flows
6 Aug 21:22:46 - [error] [api-call-service:Set LED] Call-Service attempted without connection to server.
6 Aug 21:22:46 - [error] [api-call-service:Set LED] Call-Service attempted without connection to server.
6 Aug 21:22:46 - [error] [api-call-service:Set LED] Call-Service attempted without connection to server.

Those LED messages are using the Z-Wave API so I assume my connection to HA gets dropped immediately.

This is probably a super dumb question, but how do I find the add-on repo? I thought this was the right place but it's a bit confusing since there seem to be several pieces to this puzzle.

Thanks for responding!

**Reply by lowgatelabs** [reply]
I've tried uninstalling/reinstalling everything, including rolling back to Node Red 2.0.2. Any other suggested routes?

To recap, when Node-Red is configured to connect without the integration using an Admin user key, I get endless lines of this in the NR log:

10 Aug 16:13:21 - [info] [server:Home Assistant] Connecting to http://homeassistant.myfabulous.net:8123

On the HA side, I get endless:

2022-08-10 16:10:32.134 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [139686102355680] Error handling message: Unauthorized (unauthorized)

The only thing I can figure is that NR has some bad authentication token somewhere, and despite deleting the server configuration, restarting, etc., I can't seem to blast it away. Any thoughts on how I can clear that? I looked into removing the web hook palette from NR but that wants me to delete all my nodes that use the palette. I did try manually deleting and reinstalling the palette via nom, but that didn't help.

Thanks for any guidance. I'm getting to the end of my rope on this one!

**Reply by zachowj** [reply]
Are you sure you only have 1 server config? Generate a  new access token in HA and try that.

**Reply by lowgatelabs** [reply]
I found the global context with HA configuration and there was a bunch of junk in there. I deleted it, restarted NR and HA, generated a new token and same "unauthorized" symptoms. The NR configuration doesn't seem to get any data as "states" and "services" are both empty, and isConnected and isRunning are both false.

I'm getting over my skis, but I assume that NR isn't even getting an initial "handshake" with the custom component. Is there cached info somewhere in the custom component that I should blow away? I've tried several permutations of deleting, rebooting, and reinstalling the custom component but that doesn't seem to do the trick.

**Reply by zachowj** [reply]
The custom component has nothing to do with allowing NR to connect to HA.

**Reply by zachowj** [reply]
The global variables don't get populated until NR has connected and authorized with HA. 

If you are seeing NR connect to HA in the HA logs then your issue is with the access token. Are you creating that and putting it in the server config node in Node-RED?

<https://zachowj.github.io/node-red-contrib-home-assistant-websocket/guide/#generate-access-token>

**Reply by lowgatelabs** [reply]
Followed you instructions exactly and deleted all my long-lived tokens for good measure and created a new one.

NR:
10 Aug 17:15:27 - [info] [server:Home Assistant] Connecting to http://server:8123
10 Aug 17:15:32 - [info] [server:Home Assistant] Connecting to http://server:8123
10 Aug 17:15:37 - [info] [server:Home Assistant] Connecting to http://server:8123
10 Aug 17:15:42 - [info] [server:Home Assistant] Connecting to http://server:8123
10 Aug 17:15:47 - [info] [server:Home Assistant] Connecting to http://server:8123

Verified I can ping and NSLOOKUP server from the NR command line. Both are on same subnet.

HA:
2022-08-10 17:15:47.308 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [140400038103696] Error handling message: Unauthorized (unauthorized)
2022-08-10 17:16:48.314 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [140399698112768] Error handling message: Unauthorized (unauthorized)
2022-08-10 17:17:49.437 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [140399696474416] Error handling message: Unauthorized (unauthorized)
2022-08-10 17:18:50.251 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [140399476044128] Error handling message: Unauthorized (unauthorized)

**Reply by zachowj** [reply]
Can you post a screenshot of your server config?

**Reply by zachowj** [reply]
Here's a flow to test your websocket connection Home Assistant.
1. Edit the "enter access token here" node and paste in an access token.
2. deploy
3. In the debug window you should see 
![image](https://user-images.githubusercontent.com/37859597/184034702-4e78845b-9a33-49a3-8414-5968a39c9c53.png)
4. After the `auth_ok` is received inject the "get config" and it should respond with your server config

![image](https://user-images.githubusercontent.com/37859597/184035004-a4fd2c38-d695-4d30-9dfc-4a7db762c7ef.png)

```
[{"id":"26205ec1423be98e","type":"websocket in","z":"1fd614b871996810","name":"","server":"","client":"06e5bb654365913d","x":228,"y":208,"wires":[["329a59e8437373bc"]]},{"id":"f10acb738a5b6496","type":"websocket out","z":"1fd614b871996810","name":"","server":"","client":"06e5bb654365913d","x":1178,"y":208,"wires":[]},{"id":"899fe2e93679dd5a","type":"switch","z":"1fd614b871996810","name":"","property":"payload.type","propertyType":"msg","rules":[{"t":"eq","v":"auth_required","vt":"str"},{"t":"else"}],"checkall":"true","repair":false,"outputs":2,"x":566,"y":208,"wires":[["d3cd01a673e31966"],["629b93d91a6abd9f"]]},{"id":"329a59e8437373bc","type":"json","z":"1fd614b871996810","name":"","property":"payload","action":"obj","pretty":false,"x":442,"y":208,"wires":[["899fe2e93679dd5a"]]},{"id":"d3cd01a673e31966","type":"change","z":"1fd614b871996810","name":"enter access token here","rules":[{"t":"delete","p":"_session","pt":"msg"},{"t":"set","p":"access_token","pt":"msg","to":"ENTER TOKEN HERE","tot":"str"},{"t":"set","p":"payload","pt":"msg","to":"{\"type\": \"auth\", \"access_token\": access_token}","tot":"jsonata"}],"action":"","property":"","from":"","to":"","reg":false,"x":758,"y":208,"wires":[["fa75f5242683a341","c5d878eecf5fea16"]]},{"id":"fa75f5242683a341","type":"json","z":"1fd614b871996810","name":"","property":"payload","action":"str","pretty":false,"x":950,"y":208,"wires":[["f10acb738a5b6496"]]},{"id":"864dc2b78dfb0b8a","type":"inject","z":"1fd614b871996810","name":"get config","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"{\"type\": \"get_config\"}","payloadType":"jsonata","x":644,"y":160,"wires":[["48fde12a96b71962"]]},{"id":"629b93d91a6abd9f","type":"debug","z":"1fd614b871996810","name":"debug 14","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"payload","targetType":"msg","statusVal":"","statusType":"auto","x":708,"y":256,"wires":[]},{"id":"48fde12a96b71962","type":"function","z":"1fd614b871996810","name":"set id","func":"let id = context.get(\"id\") ?? 1;\nmsg.payload = Object.assign(msg.payload, {id: id++});\ncontext.set(\"id\", id);\n\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":794,"y":160,"wires":[["fa75f5242683a341"]]},{"id":"c5d878eecf5fea16","type":"debug","z":"1fd614b871996810","name":"debug 15","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"payload","targetType":"msg","statusVal":"","statusType":"auto","x":956,"y":160,"wires":[]},{"id":"06e5bb654365913d","type":"websocket-client","path":"ws://server:8123/api/websocket","tls":"","wholemsg":"false","hb":"0","subprotocol":""}]
```
